### DISCUSSION TITLE
What's the best way to test a function with thousands of values, and also get the inputs to display when it fails?

### DISCUSSION THREAD

**Discussion by SimonEast** [vitest-dev/vitest#6385] [Q&A]
I'm trying to push many thousands of values through a test function to check for any possible math logic errors. I'd like to know:

1. Whether this is the correct way to do it
2. When an iteration fails, how can I display the function's input value, output value, and expected result

Here's a very simplifed example showing what I'm trying to achieve:

```js
// Test all values between 0 and 1 million to check for the correct answer
test('times2 function', () => {
    for (let i = 0; i < 1000000; i++) {
        expect(times2(i)).toBe(i * 2);
    }
});
```

I don't want to use `test.each()` in this case because it would create *way* too many tests. So I need to do the loop inside the test. 

But that also means that I cannot see the value of `i` when a failure is triggered. It will show the output of `times2()` and the expected value, but not `i` itself. Is there a clean way to get Vitest to tell me the value of `i` or do I need to write a custom matcher with `expect.extend` and pass in the value of `i`? That seems a little clunky for something like this.

**Comment by SimonEast** [comment]
Actually, I came across [`assert.equal()`](https://vitest.dev/api/assert.html) which has an extra parameter where you can pass a custom error message. Nice!

I typically use the `expect()` API, but perhaps I need to use the `assert` API in this case as I could then define a custom a message that displays `i` (and any other parameters if I needed multiple). 

For example:

```js
// Pass in i as the custom error message
assert.equal(times2(i), i * 2, i)
```

Or if there are multiple parameters:

```js
assert.equal(times2(i, j, k), i * j * k, [i, j, k])
```

Unless anyone has a better suggestion?

**Reply by hi-ogawa** [reply]
Using `assert` sounds okay to me but it looks like we also have an undocumented feature to pass message to `expect` https://github.com/vitest-dev/vitest/issues/6337

**Reply by SimonEast** [reply]
Oooh, I did not know about that! Will give that a try. Thank you! 😊👏
